4_basic_assembly/2_bitwise_operations/2_practical_bit_games/answers/3_write_code/3.asm

```diff
@@ -0,0 +1,87 @@
+; Basic Assembly
+; ==============
+;
+; 3.  Sum of distances:
+;    
+;     For a binary number x, we define the "sum of 1 distances" to be the sum of
+;     distances between every two "1" bits in x's binary representation.
+;
+;     Small example (8 bits):
+; 
+;       x = {10010100}_2
+;            7  4 2
+;      
+;       The total sum of distances: (7-4) + (7-2) + (4-2) = 3 + 5 + 2 = 10
+;
+;     Create a program that takes a number x (of size 4 bytes) as input, and
+;     outputs x's "sum of 1 distances".
+;
+; Based on SO: 
+; https://stackoverflow.com/questions/32470046/calculating-sum-of-all-the-distance/32470309#32470309
+
+format PE console
+entry start
+
+include 'win32a.inc' 
+
+; ===============================================
+section '.text' code readable executable
+
+start:
+
+    call    read_hex
+    mov     ecx,32d
+    mov     edx,0
+
+; count the number of ones in a number
+sum_bits:
+    mov     ebx,eax
+    and     ebx,1
+    add     edx,ebx
+    ror     eax,1
+    loop    sum_bits
+
+; calculate the sum of all 1 distances
+    mov     ebx,edx
+    mov     ecx,32d
+    mov     edx,0
+    mov     esi,0
+    mov     edi,0
+
+rotate:
+    shl     eax,1
+    jc      position
+    loop    rotate
+    jmp     exit
+
+position:
+    cmp     edx,0
+    jz      start_one
+    sub     edx,ecx         ; calculate distance
+    jmp     calc
+
+start_one:
+    mov     edx,ecx         ; move the first one position to edx
+    loop    rotate
+
+; calculate i*(n-i)*distance
+; i ... index (esi), n ... sum of all ones, distances respectively to i
+calc:
+    inc     esi
+    dec     ebx
+    jz      exit
+    imul    edx,esi
+    imul    edx,ebx

```

Comment:
@liara : I think that only one multiplication is needed. I admit that I did not try to give inputs, but two multiplications seem strange to me. Did you try to give the example `{10010100}_2` (See the top of the source file) and see if you get 10?

Reply:
I did try. The results are correct for me (see image), if I understand this assignment correctly. I get the same results if I calculate by hand. Did you check the SO link in 3.asm? Will see, If I can think of something else.

![3asm](https://user-images.githubusercontent.com/4127869/42725621-20af3174-8787-11e8-9e2f-573028a95362.PNG)

Reply:
@liara : This is very interesting! I can not argue with code that prints the correct results, I will go back to read the code. Thank you for this solution!